### PR TITLE
Fix subscription disconnection because of race conditions

### PR DESCRIPTION
**Pull request by Kunal-Jha**
# One-line summary

> Zalando ticket : ARUHA-1996 (only if appropriate)

## Description
Added a Warning message and retires for 101 status code

### CODE REVIEW

File: src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java [hunk 1]
```diff
@@ -252,12 +246,23 @@ public final void unregisterSession(final Session session) {
...
-                this::deserializeSession
-        ).values();
+        Map result = new HashMap();
+        for (int i = 0; i < 3; i++) {
+            result = loadDataAsync(zkSessions, key -> getSubscriptionPath("/sessions/" + key),
+                    this::deserializeSession);
+            if (result.size() == zkSessions.size())
+                return result.values();
+        }
+        if (result.size() != zkSessions.size()) {
```
Comment (antban) [inline src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java]: This condition is not needed

File: src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java [hunk 2]
```diff
@@ -252,12 +246,23 @@ public final void unregisterSession(final Session session) {
         } catch (Exception ex) {
             throw new NakadiRuntimeException(ex);
         }
-
-        return loadDataAsync(
-                zkSessions,
-                key -> getSubscriptionPath("/sessions/" + key),
-                this::deserializeSession
-        ).values();
+        Map result = new HashMap();
```
Comment (antban) [inline src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java]: It makes sense to use generics

File: src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java [hunk 3]
```diff
@@ -252,12 +246,23 @@ public final void unregisterSession(final Session session) {
...
+        for (int i = 0; i < 3; i++) {
+            result = loadDataAsync(zkSessions, key -> getSubscriptionPath("/sessions/" + key),
+                    this::deserializeSession);
+            if (result.size() == zkSessions.size())
```
Comment (antban) [inline src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java]: We are always using `{` and `}` to define code blocks

File: src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java [hunk 4]
```diff
@@ -252,12 +246,23 @@ public final void unregisterSession(final Session session) {
...
             throw new NakadiRuntimeException(ex);
         }
-
-        return loadDataAsync(
-                zkSessions,
-                key -> getSubscriptionPath("/sessions/" + key),
-                this::deserializeSession
-        ).values();
+        Map result = new HashMap();
+        for (int i = 0; i < 3; i++) {
```
Comment (antban) [inline src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java]: It makes no sense to query same sessions all the time - if they are changed, we have to get the new data from zk (line 243)

**Comment by antban** [comment]
:+1:

**Comment by antban** [comment]
:+1:

**Comment by Kunal-Jha** [comment]
👍

**Pull request merged**